File: app/src/main/java/com/renyu/androidcommonlibrary/utils/BaseObserver3.java
package com.renyu.androidcommonlibrary.utils;

import android.text.TextUtils;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;

import com.blankj.utilcode.util.ToastUtils;
import com.renyu.commonlibrary.dialog.NetworkLoadingDialog;
import com.renyu.commonlibrary.network.other.ResourceCoroutine;

public abstract class BaseObserver3<T> implements Observer<ResourceCoroutine<T>> {
    public abstract void onError(ResourceCoroutine<T> tResource);

    public abstract void onSucess(ResourceCoroutine<T> tResource);

    private AppCompatActivity activity;
    private String loadingText;

    private NetworkLoadingDialog networkLoadingDialog;

    public BaseObserver3() {

    }

    public BaseObserver3(AppCompatActivity activity) {
        this.activity = activity;
        this.loadingText = "正在加载中...";
    }

    public BaseObserver3(AppCompatActivity activity, String loadingText) {
        this.activity = activity;
        this.loadingText = loadingText;
    }

    @Override
    public void onChanged(@Nullable ResourceCoroutine<T> tResource) {
        if (tResource != null) {
            switch (tResource.getStatus()) {
                case ERROR:
                    if (networkLoadingDialog != null) {
                        networkLoadingDialog.close();
                    }
                    if (tResource.getException().getResult() == -1) {
                        ToastUtils.showShort("网络数据异常,请稍后再试");
                    } else {
                        ToastUtils.showShort(tResource.getException().getMessage());
                    }
                    onError(tResource);
                    break;
                case SUCESS:
                    if (networkLoadingDialog != null) {
                        networkLoadingDialog.close();
                    }
                    onSucess(tResource);
                    break;
                case LOADING:
                    if (activity != null && networkLoadingDialog == null) {
                        networkLoadingDialog = TextUtils.isEmpty(loadingText) ? NetworkLoadingDialog.getInstance() : NetworkLoadingDialog.getInstance(loadingText);
                        networkLoadingDialog.setDialogDismissListener(() -> networkLoadingDialog = null);
                        try {
                            networkLoadingDialog.show(activity);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                    break;
            }
        }
    }
}

File: app/src/main/java/com/renyu/androidcommonlibrary/utils/BaseObserver3.kt
package com.renyu.androidcommonlibrary.utils

import androidx.lifecycle.Observer
import com.blankj.utilcode.util.ToastUtils
import com.renyu.commonlibrary.network.other.Resource
import com.renyu.commonlibrary.network.other.Status

class BaseObserver3<T>(listenerBuilder: BaseObserver3<T>.ListenerBuilder.() -> Unit) :
    Observer<Resource<T>> {
    private val listener = ListenerBuilder().also(listenerBuilder)

    override fun onChanged(tResource: Resource<T>?) {
        if (tResource != null) {
            when (tResource.status) {
                Status.ERROR -> {
                    if (tResource.exception!!.result == -1) {
                        ToastUtils.showShort("网络数据异常,请稍后再试")
                    } else {
                        ToastUtils.showShort(tResource.exception!!.message)
                    }
                    listener.onErrorAction!!.invoke(tResource)
                }
                Status.SUCESS -> {
                    listener.onSuccessAction!!.invoke(tResource)
                }
                Status.LOADING -> {
                    listener.onLoadingAction!!.invoke()
                }
            }
        }
    }

    inner class ListenerBuilder {
        var onSuccessAction: ((Resource<T>) -> Unit)? = null
        var onErrorAction: ((Resource<T>) -> Unit)? = null
        var onLoadingAction: (() -> Unit)? = null

        fun onSuccess(onSuccessAction: ((Resource<T>) -> Unit)) {
            this.onSuccessAction = onSuccessAction
        }

        fun onError(onErrorAction: ((Resource<T>) -> Unit)) {
            this.onErrorAction = onErrorAction
        }

        fun onLoading(onLoadingAction: () -> Unit) {
            this.onLoadingAction = onLoadingAction
        }
    }
}